### PR TITLE
Proposal to split api.py into different files

### PR DESCRIPTION
**Pull request by totycro**
# Overview
 
This is a proposal of how api.py could be split up and is meant as a basis for discussion. The goal is to have a file structure like this:

```
pygeoapi/api/
├── __init__.py
└── coverages.py
└── features.py
└── processes.py
└── ...
```

This PR implements this for 2 endpoints of processes, but can be applied mechanically to all others.

There are also more possibilities for refactorings of the apis like extractions of common code or consolidation of error handling, but the idea here is to split up the file first, because it will necessarily be a huge diff, and smaller improvements can then be implemented in the single files at a later point.

One issue with the split is that `API` is a class and thus has to be in one file. This proposal changes the methods into regular functions which receive the `API` object as first parameter. Since we don't really use OOP features here, this is a simple change.

This commit performs the necessary adaption to transform an api function to the new style and shows that the changes are minor: https://github.com/geopython/pygeoapi/commit/78d6a0aef20cd67dd49ca91047b6be16af7d644d#diff-c9492e9767224068f9eb894b4558aea84f02c491d7a8975ae936f99ea7fc8232

Alternatives to this approach would be to either create a mixin in each file or to have classes for each api inheriting from a base API. I think both are more complicated but I'm happy to discuss 

This proposal also includes an adapter from api functions to web framework endpoints (only implemented for flask now), which can replace decorators like `pre_process` and `gzip`, and which could be used in the future for a more convenient error handling or other abstractions: https://github.com/geopython/pygeoapi/commit/78d6a0aef20cd67dd49ca91047b6be16af7d644d#diff-631fe735514040ce51f87cfb0c8eeeb1b7895d42c9364cea597355062732c159R407
This proposal would also work without this adapter if we apply the decorators to all api functions as before.

Additionally to splitting up api.py, also the tests are moved to according files (https://github.com/geopython/pygeoapi/commit/f452d454d8736978889a21ca6b3c26d711a5dbb3) as well as the openapi defintion (https://github.com/geopython/pygeoapi/commit/b7d9201c9294acb05125415285cce3c19e9a3bb0)

Once we agree on a style for this whole refactoring, I can apply it to the whole api.py file.
 

# Contributions and Licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [X] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution.
- [X] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by totycro** [comment]
I just rebased the branch. The tests currently fail due to a python version issue, which i can fix if we decide to actually follow this approach.

File: pygeoapi/api/__init__.py
Comment (francbartoli) [inline pygeoapi/api/__init__.py:381]: We should adopt a `Protocol` for the incoming request from Web Frameworks and also adopt the dependency injection pattern to pass it to the `APIRequest ` class

**Review by francbartoli — CHANGES_REQUESTED** [review]

**Comment by tomkralidis** [comment]
Strategy:
- split `pygeoapi/api.py` into one file per API standard/specification, so:
  - `pygeoapi/api/__init__.py`
  - `pygeoapi/api/processes.py`
  - `pygeoapi/api/features.py`
  - `pygeoapi/api/coverages.py`
  - `pygeoapi/api/records.py`
  - `pygeoapi/api/environmental_data_retrieval.py`
  - `pygeoapi/api/maps.py`
  - `pygeoapi/api/tiles.py` 
  - `pygeoapi/api/stac.py`

@totycro will begin with processes as the model to follow.

The above needs to work against existing tests for parity, at which point we can then refactor `tests/test_api.py` in the same manner.

**Comment by tomkralidis** [comment]
Prior to starting this work, we will wait for the approve/merge of:
- #1313 
- #1311

**Comment by tomkralidis** [comment]
Updates:
- initial refactoring complete
- next steps
  - refactor `tests/test_api.py` using the same design pattern (per API)
  - ensure all tests pass
  - perform manual testing (for UI)
  - fix merge conflicts
    - if not possible, we will send a note to the pygeoapi mailing list asking those who are `git clone`d against pygeoapi master branch to hard reset
  - circulate for final review by pygeoapi devs
- merge!

**Comment by tomkralidis** [comment]
FYI this PR is now ready for review.  I'd like to have this merged by **Friday 05 April 2024**, so review by that time would be valuable/appreciated.

**Review by GeoSander — APPROVED** [review]
This is a good start (and brave venture 😉) @totycro!

I don't see any major downsides really, accept for some increased verbosity perhaps (e.g. having to do more imports), so I would approve of the proposed changes. 👍

**Review by kalxas — APPROVED** [review]

**Review by webb-ben — APPROVED** [review]
+1

**Comment by pvgenuchten** [comment]
nice work tom, agree with sander

**Comment by GeoSander** [comment]
> nice work tom, agree with sander

To be fair, it was mostly *not* Tom's work ;)
My mistake!

File: pygeoapi/openapi.py
Comment (ricardogsilva) [inline pygeoapi/openapi.py:68]: This function's type annotation got mixed up and is now incorrect - it seemed to be correct before

File: pygeoapi/openapi.py
Comment (ricardogsilva) [inline pygeoapi/openapi.py:459]: How about having each api module's `get_oas_30()` function return a tuple of `list` and `dict` instead? It would make it more explicit than returning a dict with `tags` and `paths` keys inside of it.

So rather than:

`get_oas30() -> dict:`

it would be:

`get_oas30() -> tuple[list[str], dict[str, dict]]`

File: pygeoapi/flask_app.py
Comment (ricardogsilva) [inline pygeoapi/flask_app.py:121]: Is the completion of this TODO not in scope of this PR? I would think it is, but I'm not sure - I was going to ask about the reason for keeping and using `get_response()` in some places while using `execute_from_flask()` in others, making a similar point about having two similar but different functions as I did for the starlette app.

File: pygeoapi/django_/views.py
Comment (ricardogsilva) [inline pygeoapi/django_/views.py:538]: I guess this is the same type of comment as I left on both the starlette and flask modules - I'm not sure if this is in scope of this PR but it seems to me that it should be.

Leaving it like this implies that we keep some pretty similar functions around and use them in similar places, which will likely cause confusion and will introduce subtle bugs. I would be in favor of having this PR completely refactor the code so as to not use the older implementations and remove them.

**Review by ricardogsilva — CHANGES_REQUESTED** [review]
Great work! :tada: 

I've left some comments on things that I think could be changed, but I'm not a core maintainer, so take them with a grain of salt :wink:

Overall I think this would be a very good change to the code which will make it easier to keep improving things in the future.

**Review by doublebyte1 — COMMENTED** [review]
@totycro +1 for this effort! 👏🏽 Hopefully, from now on, it will be easier for people to add new standards and update existing ones. I left some comments, but they are more clarifications than anything else.

File: pygeoapi/starlette_app.py
Comment (tomkralidis) [reply pygeoapi/starlette_app.py:160]: Good idea; suggest to defer until this PR is completed/merged.

File: pygeoapi/process/manager/base.py
Comment (tomkralidis) [reply pygeoapi/process/manager/base.py:321]: Fixed.

File: pygeoapi/process/manager/base.py
Comment (tomkralidis) [reply pygeoapi/process/manager/base.py:1]: This is an artifact of a manual backport (there are a few) over the course of PR development and commits to master.

File: pygeoapi/process/manager/dummy.py
Comment (tomkralidis) [reply pygeoapi/process/manager/dummy.py:1]: This is an artifact of a manual backport (there are a few) over the course of PR development and commits to master.

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:1]: Minor fix pushed.

File: pygeoapi/openapi.py
Comment (tomkralidis) [reply pygeoapi/openapi.py:68]: Fixed

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:1]: Fixed/updated.

File: pygeoapi/openapi.py
Comment (tomkralidis) [reply pygeoapi/openapi.py:459]: Good idea! Done.

File: pygeoapi/api/coverages.py
Comment (tomkralidis) [reply pygeoapi/api/coverages.py:205]: Fixed

File: pygeoapi/api/stac.py
Comment (tomkralidis) [reply pygeoapi/api/stac.py:65]: Fixed

File: pygeoapi/api/maps.py
Comment (tomkralidis) [reply pygeoapi/api/maps.py:248]: Fixed

File: pygeoapi/api/processes.py
Comment (tomkralidis) [reply pygeoapi/api/processes.py:559]: Fixed

File: pygeoapi/api/stac.py
Comment (tomkralidis) [reply pygeoapi/api/stac.py:228]: Fixed

File: pygeoapi/api/tiles.py
Comment (tomkralidis) [reply pygeoapi/api/tiles.py:441]: Fixed

File: tests/api/__init__.py
Comment (tomkralidis) [reply tests/api/__init__.py:1]: Fixed.

File: pygeoapi/flask_app.py
Comment (tomkralidis) [reply pygeoapi/flask_app.py:121]: @totycro ^^

**Comment by tomkralidis** [comment]
Thanks for the feedback @ricardogsilva / @doublebyte1.  I've tried to address most of the issues / comments in time for @totycro's Wednesday morning, as well as suggestion to defer other suggestions following the merge of this PR.

Note that there is an unrelated fix in https://github.com/geopython/pygeoapi/pull/1405/commits/eeb997c7207a052a140e02972f196f3b625fb0ab in support of passing the EDR tests based on possible data changes.

**Review by doublebyte1 — APPROVED** [review]

File: pygeoapi/api/__init__.py
Comment (totycro) [reply pygeoapi/api/__init__.py]: How would a `Protocol` look like for this? The different request classes have some different fields, so I'm not sure how it could be combined to one `Protocol`

File: pygeoapi/api/itemtypes.py
Comment (totycro) [reply pygeoapi/api/itemtypes.py:130]: Most likely this has been ported from the old version of the code. I'd suggest to not do any behavioral changes in this PR as it already changes a lot of other things, but we could change this after the merge or open an issue to keep it in mind.

File: pygeoapi/starlette_app.py
Comment (totycro) [reply pygeoapi/starlette_app.py:160]: I also agree, this is also what the related TODOs hint towards. We could address this soon after the merge

File: pygeoapi/flask_app.py
Comment (totycro) [reply pygeoapi/flask_app.py:121]: I think this is really the same discussion as https://github.com/geopython/pygeoapi/pull/1405#discussion_r1545755291

File: pygeoapi/django_/views.py
Comment (totycro) [reply pygeoapi/django_/views.py:538]: This is also related to https://github.com/geopython/pygeoapi/pull/1405#discussion_r1545755291

**Comment by totycro** [comment]
I think I addressed all of the comments by now. The way I see it, we don't have any issues really blocking the merge, and I'd prefer to merge it ASAP because many commits done on master right now will introduce new conflicts.

File: pygeoapi/api/itemtypes.py
Comment (tomkralidis) [reply pygeoapi/api/itemtypes.py:130]: This was the result of adding `.../schema` support for coverages.  This is now moved to `pygeoapi/api/__init__.py` and called accordingly from upstream.

**Review by ricardogsilva — APPROVED** [review]
I understand it is your intention to address my remaining comments after the merge, which I would be happy with.

+1 from me, thanks for the hard work!

File: pygeoapi/openapi.py
Comment (tomkralidis) [reply pygeoapi/openapi.py:170]: Fixed

File: pygeoapi/openapi.py
Comment (tomkralidis) [reply pygeoapi/openapi.py:311]: This can be found in https://github.com/totycro/pygeoapi/blob/split-api-py/pygeoapi/openapi.py#L402

**Review by tomkralidis — APPROVED** [review]

**Pull request merged**
